File: wizards/com/sun/star/wizards/web/data/CGSettings.java
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This file incorporates work covered by the following license notice:
 *
 *   Licensed to the Apache Software Foundation (ASF) under one or more
 *   contributor license agreements. See the NOTICE file distributed
 *   with this work for additional information regarding copyright
 *   ownership. The ASF licenses this file to you under the Apache
 *   License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of
 *   the License at http://www.apache.org/licenses/LICENSE-2.0 .
 */

/*
 * IconSet.java
 *
 * Created on 12. September 2003, 17:16
 */
package com.sun.star.wizards.web.data;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import com.sun.star.beans.PropertyValue;
import com.sun.star.container.NoSuchElementException;
import com.sun.star.container.XNameAccess;
import com.sun.star.i18n.NumberFormatIndex;
import com.sun.star.lang.WrappedTargetException;
import com.sun.star.lang.XMultiServiceFactory;
import com.sun.star.uno.UnoRuntime;
import com.sun.star.util.DateTime;
import com.sun.star.wizards.common.ConfigGroup;
import com.sun.star.wizards.common.ConfigSet;
import com.sun.star.wizards.common.FileAccess;
import com.sun.star.wizards.common.Helper.DateUtils;
import com.sun.star.wizards.common.JavaTools;
import com.sun.star.wizards.common.Properties;
import com.sun.star.wizards.common.PropertyNames;

public class CGSettings extends ConfigGroup
{

    public String soTemplateDir;
    public String soGalleryDir;
    public String workPath = null;
    public String cp_WorkDir;
    public ConfigSet cp_Exporters = new ConfigSet(CGExporter.class);
    public ConfigSet cp_Layouts = new ConfigSet(CGLayout.class);
    public ConfigSet cp_Styles = new ConfigSet(CGStyle.class);
    public ConfigSet cp_IconSets = new ConfigSet(CGIconSet.class);
    public ConfigSet cp_BackgroundImages = new ConfigSet(CGImage.class);
    public ConfigSet cp_SavedSessions = new ConfigSet(CGSessionName.class);
    public ConfigSet cp_Filters = new ConfigSet(CGFilter.class);
    public ConfigSet savedSessions = new ConfigSet(CGSessionName.class);
    public CGSession cp_DefaultSession = new CGSession();
    public String cp_LastSavedSession;
    private Map<String, CGExporter[]> exportersMap = new HashMap<String, CGExporter[]>();
    private XMultiServiceFactory xmsf;
    String[] resources;
    public Formatter formatter;
    public static final int RESOURCE_PAGES_TEMPLATE = 0;
    public static final int RESOURCE_SLIDES_TEMPLATE = 1;
    public static final int RESOURCE_CREATED_TEMPLATE = 2;
    public static final int RESOURCE_UPDATED_TEMPLATE = 3;
    public static final int RESOURCE_SIZE_TEMPLATE = 4;

    public CGSettings(XMultiServiceFactory xmsf_, String[] resources_, Object document)
    {
        xmsf = xmsf_;
        try
        {
            soTemplateDir = FileAccess.getOfficePath(xmsf, "Config", PropertyNames.EMPTY_STRING, PropertyNames.EMPTY_STRING);
            soGalleryDir = FileAccess.getOfficePath(xmsf, "Gallery", "share", PropertyNames.EMPTY_STRING);
            root = this;
            formatter = new Formatter(xmsf, document);
            resources = resources_;
        }
        catch (Exception ex)
        {
            ex.printStackTrace();
        }
    }
    private static final CGExporter[] EMPTY_ARRAY_1 = new CGExporter[0];

    public CGExporter[] getExporters(String mime)
    {
        CGExporter[] exps = exportersMap.get(mime);
        if (exps == null)
        {
            exportersMap.put(mime, exps = createExporters(mime));
        }
        return exps;
    }

    private CGExporter[] createExporters(String mime)
    {
        Object[] exporters = cp_Exporters.items();
        ArrayList<Object> v = new ArrayList<Object>();
        for (int i = 0; i < exporters.length; i++)
        {
            if (((CGExporter) exporters[i]).supports(mime))
            {
                try
                {
                    v.add(exporters[i]);
                }
                catch (Exception ex)
                {
                    ex.printStackTrace();
                }
            }
        }
        return v.toArray(new CGExporter[v.size()]);
    }

    /**
     * call after read.
     * @param xmsf
     */
    public void configure(XMultiServiceFactory xmsf) throws Exception
    {
        workPath = FileAccess.connectURLs(soTemplateDir, cp_WorkDir);
        calcExportersTargetTypeNames(xmsf);
    }

    private void calcExportersTargetTypeNames(XMultiServiceFactory xmsf) throws Exception
    {
        Object typeDetect = xmsf.createInstance("com.sun.star.document.TypeDetection");
        XNameAccess xNameAccess = UnoRuntime.queryInterface(XNameAccess.class, typeDetect);
        for (int i = 0; i < cp_Exporters.getSize(); i++)
        {
            calcExporterTargetTypeName(xNameAccess, (CGExporter) cp_Exporters.getElementAt(i));
        }
    }

    private void calcExporterTargetTypeName(XNameAccess xNameAccess, CGExporter exporter)
            throws NoSuchElementException,
            WrappedTargetException
    {
        if (!exporter.cp_TargetType.equals(PropertyNames.EMPTY_STRING))
        {
            exporter.targetTypeName =
                    (String) Properties.getPropertyValue(
                    (PropertyValue[]) xNameAccess.getByName(exporter.cp_TargetType),
                    "UIName");
        }
    }
    FileAccess fileAccess;

    FileAccess getFileAccess() throws Exception
    {
        return getFileAccess(xmsf);
    }

    FileAccess getFileAccess(XMultiServiceFactory xmsf) throws Exception
    {
        if (fileAccess == null)
        {
            fileAccess = new FileAccess(xmsf);
        }
        return fileAccess;
    }

    public class Formatter
    {

        private long docNullTime;
        private int dateFormat,  numberFormat;
        private DateUtils dateUtils;

        public Formatter(XMultiServiceFactory xmsf, Object document) throws Exception
        {
            dateUtils = new DateUtils(xmsf, document);
            dateFormat = dateUtils.getFormat(NumberFormatIndex.DATE_SYS_DMMMYYYY);
            numberFormat = dateUtils.getFormat(NumberFormatIndex.NUMBER_1000DEC2);
        }

        public String formatCreated(DateTime date)
        {
            String sDate = dateUtils.format(dateFormat, date);
            return JavaTools.replaceSubString(resources[CGSettings.RESOURCE_CREATED_TEMPLATE], sDate, "%DATE");
        }

        public String formatUpdated(DateTime date)
        {
            String sDate = dateUtils.format(dateFormat, date);
            return JavaTools.replaceSubString(resources[CGSettings.RESOURCE_UPDATED_TEMPLATE], sDate, "%DATE");
        }

        public String formatFileSize(int size)
        {
            float sizeInKB = size / 1024f;
            String sSize = dateUtils.getFormatter().convertNumberToString(numberFormat, sizeInKB);
            return JavaTools.replaceSubString(resources[CGSettings.RESOURCE_SIZE_TEMPLATE], sSize, "%NUMBER");
        }
    }
}

File: wizards/com/sun/star/wizards/web/data/CGSettings.py
#
# This file is part of the LibreOffice project.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
# This file incorporates work covered by the following license notice:
#
#   Licensed to the Apache Software Foundation (ASF) under one or more
#   contributor license agreements. See the NOTICE file distributed
#   with this work for additional information regarding copyright
#   ownership. The ASF licenses this file to you under the Apache
#   License, Version 2.0 (the "License"); you may not use this file
#   except in compliance with the License. You may obtain a copy of
#   the License at http://www.apache.org/licenses/LICENSE-2.0 .
#
import traceback
from datetime import date as dateTimeObject

from ...common.FileAccess import FileAccess
from ...common.ConfigGroup import ConfigGroup
from ...common.NumberFormatter import NumberFormatter
from ...common.Properties import Properties
from ..WebConfigSet import WebConfigSet
from .CGExporter import CGExporter
from .CGLayout import CGLayout
from .CGStyle import CGStyle
from .CGIconSet import CGIconSet
from .CGImage import CGImage
from .CGFilter import CGFilter
from .CGSessionName import CGSessionName
from .CGSession import CGSession

from com.sun.star.i18n.NumberFormatIndex import DATE_SYS_DMMMYYYY
from com.sun.star.i18n.NumberFormatIndex import NUMBER_1000DEC2

class CGSettings(ConfigGroup):

    RESOURCE_PAGES_TEMPLATE = 0
    RESOURCE_SLIDES_TEMPLATE = 1
    RESOURCE_CREATED_TEMPLATE = 2
    RESOURCE_UPDATED_TEMPLATE = 3
    RESOURCE_SIZE_TEMPLATE = 4

    def __init__(self, xmsf_, resources_, document):
        self.cp_WorkDir = str()
        self.cp_Exporters = WebConfigSet(CGExporter)
        self.cp_Layouts = WebConfigSet(CGLayout)
        self.cp_Styles = WebConfigSet(CGStyle)
        self.cp_IconSets = WebConfigSet(CGIconSet)
        self.cp_BackgroundImages = WebConfigSet(CGImage)
        self.cp_SavedSessions = WebConfigSet(CGSessionName)
        self.cp_Filters = WebConfigSet(CGFilter)
        self.savedSessions = WebConfigSet(CGSessionName)
        self.cp_DefaultSession = CGSession()
        self.cp_LastSavedSession = str()
        self.fileAccess = None
        self.workPath = None
        self.xmsf = xmsf_
        try:
            self.soTemplateDir = FileAccess.getOfficePath2(self.xmsf, "Config", "", "");
            self.soGalleryDir = FileAccess.getOfficePath2(self.xmsf, "Gallery", "share", "");
            self.setRoot(self)
            self.formatter = self.Formatter(self.xmsf, document, resources_)
            self.resources = resources_
            self.exportersMap = {}
        except Exception:
            traceback.print_exc()

    def getExporters(self, mime):
        exps = self.exportersMap.get(mime)
        if exps is None:
            exps = self.createExporters(mime)
            self.exportersMap[mime] = exps
        return exps

    def createExporters(self, mime):
        exporters = self.cp_Exporters.childrenList
        v = []
        for i in exporters:
            if i.supports(mime):
                try:
                    v.append(i)
                except Exception:
                    traceback.print_exc()
        return v

    '''
    call after read.
    @param xmsf
    @param document the background document. used for date/number formatting.
    '''

    def configure(self, xmsf):
        self.workPath = FileAccess.connectURLs(self.soTemplateDir, self.cp_WorkDir)
        self.calcExportersTargetTypeNames(xmsf)

    def calcExportersTargetTypeNames(self, xmsf):
        typeDetect = xmsf.createInstance("com.sun.star.document.TypeDetection")
        for i in range(self.cp_Exporters.getSize()):
            self.calcExporterTargetTypeName(typeDetect, self.cp_Exporters.getElementAt(i))

    def calcExporterTargetTypeName(self, typeDetect, exporter):
        if (exporter is None):
            return
        if not (exporter.cp_TargetType == "" or exporter.cp_TargetType is None):
            targetTypeName = Properties.getPropertyValue(
                typeDetect.getByName(exporter.cp_TargetType), "UIName")
            exporter.cp_targetTypeName = targetTypeName

    def getFileAccess(self, xmsf = None):
        if xmsf is None:
            xmsf = self.xmsf
        if self.fileAccess is None:
            self.fileAccess = FileAccess(xmsf)
        return self.fileAccess

    class Formatter(object):
        class DateUtils(object):

            def __init__(self, xmsf, document):
                self.formatSupplier = document
                formatSettings = self.formatSupplier.getNumberFormatSettings()
                date = formatSettings.NullDate
                self.calendar = dateTimeObject(date.Year, date.Month, date.Day)
                self.formatter = NumberFormatter.createNumberFormatter(xmsf, self.formatSupplier)

            '''
            @param format a constant of the enumeration NumberFormatIndex
            @return
            '''
            def getFormat(self, f):
                return NumberFormatter.getNumberFormatterKey(self.formatSupplier, f)

            '''
            @param date a VCL date in form of 20041231
            @return a document relative date
            '''
            def format(self, formatIndex, date):
                dateTime = dateTimeObject(date.Year, date.Month, date.Day)
                difference =  dateTime - self.calendar
                return self.formatter.convertNumberToString(formatIndex,  difference.days)

            def getFormatter(self):
                return self.formatter

        def __init__(self, xmsf, document, resources):
            self.dateUtils = self.DateUtils(xmsf, document)
            self.dateFormat = self.dateUtils.getFormat(DATE_SYS_DMMMYYYY)
            self.numberFormat = self.dateUtils.getFormat(NUMBER_1000DEC2)
            self.resources = resources

        def formatCreated(self, date):
            sDate = self.dateUtils.format(self.dateFormat, date)
            return self.resources[CGSettings.RESOURCE_CREATED_TEMPLATE].replace(
                "%DATE", sDate)

        def formatUpdated(self, date):
            sDate = self.dateUtils.format(self.dateFormat, date);
            return self.resources[CGSettings.RESOURCE_UPDATED_TEMPLATE].replace(
                "%DATE", sDate)

        def formatFileSize(self, size):
            sizeInKB = size / float(1024)
            sSize = self.dateUtils.getFormatter().convertNumberToString(
                self.numberFormat, sizeInKB)
            return self.resources[CGSettings.RESOURCE_SIZE_TEMPLATE].replace(
                "%NUMBER", sSize)
